contracts/reverse-registrar/src/contract.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{
+    errors::Error,
+    events::emit_domain_updated,
+    registry::{ContractErrors as RegistryErrors, Record, RecordKeys},
+    types::{CoreDataKeys, Domain},
+};
+use common::utils::generate_node;
+use soroban_sdk::{
+    contract, contractimpl, panic_with_error, symbol_short, Address, BytesN, Env, IntoVal,
+};
+
+const LEDGER_DAY: u32 = 17_280;
+
+pub trait ReverseRegistrarTrait {
+    /// Initialize the contract with the admin and registry addresses.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - The environment in which the contract is executed.
+    /// * `admin` - The address of the admin.
+    /// * `registry` - The address of the registry.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<(), Error>` - Returns an error if the contract is already initialized.
+    fn init(e: Env, admin: Address, registry: Address) -> Result<(), Error>;
+
+    /// Upgrade the contract to a new version.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - The environment in which the contract is executed.
+    /// * `hash` - The hash of the new contract.
+    fn upgrade(e: Env, hash: BytesN<32>);
+
+    /// Set the reverse domain for an address.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - The environment in which the contract is executed.
+    /// * `address` - The address for which the reverse domain is being set.
+    /// * `domain` - The reverse domain to set, or `None` to remove the reverse domain.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<(), Error>` - Returns an error if the operation fails.
+    fn set(e: Env, address: Address, domain: Option<Domain>) -> Result<(), Error>;
+
+    /// Get the reverse domain for an address.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - The environment in which the contract is executed.
+    /// * `address` - The address for which the reverse domain is being retrieved.
+    ///
+    /// # Returns
+    ///
+    /// * `Option<Domain>` - The reverse domain if it exists, otherwise `None`.
+    fn get(e: Env, address: Address) -> Option<Domain>;
+}
+
+#[contract]
+pub struct ReverseRegistrar;
+
+#[contractimpl]
+impl ReverseRegistrarTrait for ReverseRegistrar {
+    fn init(e: Env, admin: Address, registry: Address) -> Result<(), Error> {
+        if e.storage().instance().has(&CoreDataKeys::Admin) {
+            return Err(Error::AlreadyInitialized);
+        }
+
+        e.storage().instance().set(&CoreDataKeys::Admin, &admin);
+        e.storage()
+            .instance()
+            .set(&CoreDataKeys::Registry, &registry);
+        Ok(())
+    }
+
+    fn upgrade(e: Env, hash: BytesN<32>) {
+        e.storage()
+            .instance()
+            .get::<CoreDataKeys, Address>(&CoreDataKeys::Admin)
+            .unwrap()
+            .require_auth();
+
+        e.deployer().update_current_contract_wasm(hash);
+    }
+
+    fn set(e: Env, address: Address, domain: Option<Domain>) -> Result<(), Error> {
```

Comment:
Similar to the KeyValue contract, when setting a new record the one setting a new value should pay a fee. On Mainnet that fee will be paid in $PSD

Reply:
Thank you for pointing this out. I have added it, and when adding or modifying it, the user will need to pay a fee, but removing it does not require payment.